File: jas_ocaml/test/algorithms/pattern_along_path_test.ml
(* Equivalence vectors shared with the Rust reference
   (jas_dioxus/src/algorithms/pattern_along_path.rs) and the Swift port: a
   diamond side tile tiled twice along a straight 100-long path. *)

let brush () : Jas.Pattern_along_path.t =
  Jas.Pattern_along_path.{
    tile_width = 100.0;
    tile_height = 20.0;
    side = [ [ (0.0, 10.0); (50.0, 0.0); (100.0, 10.0); (50.0, 20.0) ] ];
    scale = 100.0;
    spacing = 0.0;
    flip_across = false;
    flip_along = false;
    stroke_weight = 10.0;
  }

let close (ax, ay) (bx, by) =
  Float.abs (ax -. bx) < 1e-6 && Float.abs (ay -. by) < 1e-6

let test_two_tiles () =
  let cmds = [ Jas.Element.MoveTo (0.0, 0.0); Jas.Element.LineTo (100.0, 0.0) ] in
  let out = Jas.Pattern_along_path.tile cmds (brush ()) in
  Alcotest.(check int) "two tiles" 2 (List.length out);
  let t0 = Array.of_list (List.nth out 0) in
  let t1 = Array.of_list (List.nth out 1) in
  Alcotest.(check bool) "t0 start" true (close t0.(0) (0.0, 0.0));
  Alcotest.(check bool) "t0 mid-top" true (close t0.(1) (25.0, -5.0));
  Alcotest.(check bool) "t0 end" true (close t0.(2) (50.0, 0.0));
  Alcotest.(check bool) "t1 start" true (close t1.(0) (50.0, 0.0));
  Alcotest.(check bool) "t1 end" true (close t1.(2) (100.0, 0.0))

let test_degenerate () =
  let out = Jas.Pattern_along_path.tile [ Jas.Element.MoveTo (0.0, 0.0) ] (brush ()) in
  Alcotest.(check bool) "empty" true (out = [])

let () =
  Alcotest.run "PatternAlongPath"
    [
      ( "tile",
        [
          Alcotest.test_case "two tiles" `Quick test_two_tiles;
          Alcotest.test_case "degenerate" `Quick test_degenerate;
        ] );
    ]

File: jas/algorithms/pattern_along_path_test.py
"""Equivalence vectors shared with the Rust reference
(jas_dioxus/src/algorithms/pattern_along_path.rs) and the Swift / OCaml
ports: a diamond side tile tiled twice along a straight 100-long path."""

from algorithms.pattern_along_path import PatternBrush, pattern_along_path
from geometry.element import MoveTo, LineTo


def _brush():
    return PatternBrush(
        tile_width=100.0, tile_height=20.0,
        side=[[(0.0, 10.0), (50.0, 0.0), (100.0, 10.0), (50.0, 20.0)]],
        scale=100.0, spacing=0.0, flip_across=False, flip_along=False,
        stroke_weight=10.0,
    )


def _close(a, x, y):
    return abs(a[0] - x) < 1e-6 and abs(a[1] - y) < 1e-6


def test_straight_path_tiles_twice():
    out = pattern_along_path([MoveTo(0.0, 0.0), LineTo(100.0, 0.0)], _brush())
    assert len(out) == 2
    assert _close(out[0][0], 0.0, 0.0)
    assert _close(out[0][1], 25.0, -5.0)
    assert _close(out[0][2], 50.0, 0.0)
    assert _close(out[1][0], 50.0, 0.0)
    assert _close(out[1][2], 100.0, 0.0)


def test_empty_for_degenerate():
    assert pattern_along_path([MoveTo(0.0, 0.0)], _brush()) == []
